src/test/java/org/pkcs11/jacknji11/CryptokiTest.java

```diff
@@ -265,53 +265,164 @@ public void testDigest() {
+        }
+    }
+
+    public void testSignVerifyECDSA() {
+        long session = CE.OpenSession(TESTSLOT, CK_SESSION_INFO.CKF_RW_SESSION | CK_SESSION_INFO.CKF_SERIAL_SESSION, null, null);
+        CE.LoginUser(session, USER_PIN);
+        // Attributes from PKCS #11 Cryptographic Token Interface Current Mechanisms Specification Version 2.40 section 2.3.3 - ECDSA public key objects
+        /* DER-encoding of an ANSI X9.62 Parameters, also known as "EC domain parameters". */
+        // *Comment* See X9.62-1998 Public Key Cryptography For The Financial Services Industry: The Elliptic Curve Digital Signature Algorithm (ECDSA)
+        // page 27.
+        // We use a P-256 key (also known as secp256r1 or prime256v1), the oid 1.2.840.10045.3.1.7 has DER encoding in Hex 06082a8648ce3d030107
+        byte[] ecCurveParams = Hex.s2b("06082a8648ce3d030107");
+        CKA[] pubTempl = new CKA[] {
+            new CKA(CKA.EC_PARAMS, ecCurveParams),
+            new CKA(CKA.WRAP, false),
+            new CKA(CKA.ENCRYPT, false),
+            new CKA(CKA.VERIFY, true),
+            new CKA(CKA.VERIFY_RECOVER, false),
+            new CKA(CKA.TOKEN, true),
+            new CKA(CKA.LABEL, "label-public"),
+            new CKA(CKA.ID, "label"),
+        };
+        CKA[] privTempl = new CKA[] {
+            new CKA(CKA.TOKEN, true),
+            new CKA(CKA.PRIVATE, true),
+            new CKA(CKA.SENSITIVE, true),
+            new CKA(CKA.SIGN, true),
+            new CKA(CKA.SIGN_RECOVER, false),
+            new CKA(CKA.DECRYPT, false),
+            new CKA(CKA.UNWRAP, false),
+            new CKA(CKA.EXTRACTABLE, false),
+            new CKA(CKA.LABEL, "label-private"),
+            new CKA(CKA.ID, "label"),
+        };
+        LongRef pubKey = new LongRef();
+        LongRef privKey = new LongRef();
+        CE.GenerateKeyPair(session, new CKM(CKM.ECDSA_KEY_PAIR_GEN), pubTempl, privTempl, pubKey, privKey);
+
+        // Direct sign, PKCS#11 "2.3.6 ECDSA without hashing"
+        byte[] data = new byte[32]; // SHA256 hash is 32 bytes
+        CE.SignInit(session, new CKM(CKM.ECDSA), privKey.value());
+        byte[] sig1 = CE.Sign(session, data);
+        assertEquals(64, sig1.length);
+
+        CE.VerifyInit(session, new CKM(CKM.ECDSA), pubKey.value());
+        CE.Verify(session, data, sig1);
+
+        CE.VerifyInit(session, new CKM(CKM.ECDSA), pubKey.value());
+        try {
+            CE.Verify(session, data, new byte[64]);
+            fail("CE Verify with no real signature should throw exception");
+        } catch (CKRException e) {
+            assertEquals("Failure with invalid signature data should be CKR.SIGNATURE_INVALID", CKR.SIGNATURE_INVALID, e.getCKR());
+        }
+    }
+
+    /** SignRecoverInit and VerifyRecoverInit is not supported on all HSMs, so it has a separate test that may expect to fail with FUNCTION_NOT_SUPPORTED
+     */
```

Comment:
Could you wrap lines at 80/120 and format this like:

/**
 * SignRecoverIniit ...
 */

Reply:
absolutely